### PR TITLE
Introduce reasons and consolidate status conditions

### PR DESCRIPTION
**Pull request by ccremer**
## Summary

* The [API convention](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status) requires setting a status reason. Thus, lots of condition code needed updated
* Remove dependency on operator-lib

Without reasons, patching status fails with the upstream condition tools
```console
2021-01-20T11:33:46.881Z	ERROR	controllers.Prune	could not patch backup conditions	{"prune": "default/prunejob-1611139800", "error": "Prune.backup.appuio.ch \"prunejob-1611139800\" is invalid: status.conditions.reason: Invalid value: \"\": status.conditions.reason in body should be at least 1 chars long"}
```

Example backup object:
```yaml
status:
  conditions:
    - lastTransitionTime: '2021-01-21T13:27:00Z'
      message: no container definitions found
      reason: NoPreBackupPodsFound
      status: 'True'
      type: PreBackupPodsReady
    - lastTransitionTime: '2021-01-21T13:27:01Z'
      message: the job 'default/backupjob-1611235620' was created
      reason: Ready
      status: 'True'
      type: Ready
    - lastTransitionTime: '2021-01-21T13:27:01Z'
      message: The job is progressing
      reason: Started
      status: 'True'
      type: Progressing
```

## Checklist

- [ ] Keep pull requests small so they can be easily reviewed.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`,
      as they show up in the changelog
- [x] Update the documentation.
- [ ] Update tests.
- [ ] Link this PR to related issues.

### CODE REVIEW

File: config/crd/apiextensions.k8s.io/v1/backup.appuio.ch_archives.yaml [hunk 1]
```diff
@@ -332,24 +332,44 @@ spec:
...
+                  description: "Condition contains details for one aspect of the current state of this API Resource. --- This struct is intended for direct use as an array at the field path .status.conditions.  For example, type FooStatus struct{     // Represents the observations of a foo's current state.     // Known .status.conditions.type are: \"Available\", \"Progressing\", and \"Degraded\"     // +patchMergeKey=type     // +patchStrategy=merge     // +listType=map     // +listMapKey=type     Conditions []metav1.Condition `json:\"conditions,omitempty\" patchStrategy:\"merge\" patchMergeKey:\"type\" protobuf:\"bytes,1,rep,name=conditions\"` \n     // other fields }"
```
Comment (cimnine) [inline config/crd/apiextensions.k8s.io/v1/backup.appuio.ch_archives.yaml:336]: I think we should provide our own comment if possible.

File: job/status.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+func (c *Config) SetConditionTrue(condition k8upv1alpha1.ConditionType, reason k8upv1alpha1.ConditionReason) {
+	c.patchConditions(metav1.ConditionTrue, reason, "", condition)
+}
+
+// SetConditionUnknownWithMessage tells the K8s controller at once that the status of the given Conditions is "Unknown"
+func (c *Config) SetConditionUnknownWithMessage(condition k8upv1alpha1.ConditionType, reason k8upv1alpha1.ConditionReason, message string, args ...interface{}) {
+	c.patchConditions(metav1.ConditionUnknown, reason, fmt.Sprintf(message, args...), condition)
+}
+
+// SetConditionFalse tells the K8s controller at once that the status of the given Condition is now "True" and
```
Comment (cimnine) [inline job/status.go:24]: CodeClimate says: `comment on exported method Config.SetConditionTrueWithMessage should be of the form "SetConditionTrueWithMessage ..."`

File: go.sum [hunk 3]
```diff
@@ -610,6 +610,7 @@ github.com/openzipkin/zipkin-go v0.2.1/go.mod h1:NaW6tEwdmWMaCDZzg8sh+IBNOxHMPnh
 github.com/openzipkin/zipkin-go v0.2.2/go.mod h1:NaW6tEwdmWMaCDZzg8sh+IBNOxHMPnhQw8ySjnjRyN4=
 github.com/operator-framework/operator-lib v0.1.0 h1:7Qy6v2ZccvCeFLWEkrGnN+U+DkaeIWp0gAZaBM9T3DI=
 github.com/operator-framework/operator-lib v0.1.0/go.mod h1:HLw61JTIEeq0YLeVf4dwYx/zt4DmLGZUVWI1y3Lf5Hg=
+github.com/operator-framework/operator-lib v0.4.0 h1:g7tGRo+FikHgFZDmRdHkOxyTv3sViI+Ujiqbfd9Tfsk=
```
Comment (cimnine) [inline go.sum:613]: Funny that v0.1.0 of the operator-lib is still around. Would `go mod tidy` erase that, perhaps?

File: job/status.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+// SetFinished sets the `c.Obj.GetStatus().Finished` property to `true`.
+// In the same call to the k8s API it also sets the Progressing condition to "False" with reason Finished.
+func (c *Config) SetFinished(namespace, name string) {
+	status := c.Obj.GetStatus()
+	status.Finished = true
+
+	meta.SetStatusCondition(&status.Conditions, metav1.Condition{
+		Type:    k8upv1alpha1.ConditionProgressing.String(),
```
Comment (cimnine) [inline job/status.go:102]: Did you mean to call it _progressing_ or did you perhaps mean to call it _processing_?

Comment (ccremer) [reply job/status.go:102]: Yes. Like the `Deployment` has a Progressing condition. Used when doing rolling updates for example.

**Comment by ccremer** [comment]
@tobru would you like to have a glance over the documentation?

**Review by tobru — APPROVED** [review]
documentation LGTM

**Review by cimnine — APPROVED** [review]
LGTM

**Pull request merged**